Mempool: persist mempoolminfee accross restarts

ismaelsadeeq opened the pull request:
Follow-up [#27622](https://github.com/bitcoin/bitcoin/pull/27622)
See –> [Comment](https://github.com/bitcoin/bitcoin/issues/27555#issuecomment-1538504504) 
Currently, there is no persistence of mempoolminfee
As mentioned in the comment

> persist `mempoolminfee` across restarts to at least tell user a plausible value to stay in the mempool for a bit

The suggestions outlined in the comment are implemented in this pull request. The `mempoolminfee` will now be persisted on shutdown and periodically in the event of an unclean shutdown.
Upon restart, the previous `mempoolminfee` value will be remembered, and any transactions with fees below the persisted mempoolminfee will be rejected from entering the mempool.

If a user passes the `-minrelaytxfee`, it will be utilized as the `mempoolminfee`, just as it was prior to this pull request. 
When reading the persisted mempoolminfee, we take the max of the persisted value and the default minimum fee of 1000 satoshis, ensuring that the remembered `mempoolminfee` value will not be less than 1000 satoshis.

ismaelsadeeq committed:
test: reset mempool minimum fee with `minrelaytxfee` option.

This commit update tests that have the assumption that mempool
minimum fee will be reset to the default value of 1000 Satoshi after a restart,
to restart with a minrelaytxfee of 1000 satoshis, which will effectively sets
the mempool minimum fee back to its default value.

The changes made in this pull request invalidate the previous assumption, necessitating
this adjustment.

ajtowns commented on src/node/mempool_persist_args.cpp:
```diff
@@ -21,3 +21,8 @@ fs::path MempoolPath(const ArgsManager& argsman)
+
+fs::path MemPoolMinFeePath(const ArgsManager& argsman)
+{
+    return argsman.GetDataDirNet() / "mempool_min_fee.dat";
```
Why add this as a separate file, rather than adding it to the end of the fee estimates data file which is already updated periodically and ignored if out of date?

Adding a method to the mempool to set the min fee based on that seems plausible? (Perhaps make it conditional on the new value being higher than the current minfee, and the mempool being less than 95% full?)

ajtowns replied:
I guess as long as it's only set shortly after startup (while the mempool is still loading?), it's probably fine?

ismaelsadeeq replied:
I updated the commits, it's now flushing to fee_estimates.dat

ismaelsadeeq commented on doc/files.md:
```diff
@@ -64,6 +64,7 @@ Subdirectory       | File(s)               | Description
 `./`               | `guisettings.ini.bak` | Backup of former [GUI settings](#gui-settings) after `-resetguisettings` option is used
 `./`               | `ip_asn.map`          | IP addresses to Autonomous System Numbers (ASNs) mapping used for bucketing of the peers; path can be specified with the `-asmap` option
 `./`               | `mempool.dat`         | Dump of the mempool's transactions
+`./`               | `mempool_min_fee.dat` | Stores the minimum fee rate of a transaction that will be accepted into the mempool
```
It appears that taking the approach of dumping to `fee_estimates.dat` and implementing a method to set `mempoolminfee` based on the persisted value is a better option.

Your perspective on persisting the mempool periodically. I understand the concern about remembering the fee estimates and `mempoolminfee` without the mempool itself. However, I would like to seek further insight into the potential consequences that could arise from this, as dumping the mempool periodically might be unappealing

willcl-ark commented on src/policy/fees.h:
```diff
@@ -14,12 +14,25 @@
 #include <vector>
 
+
+// How often to flush fee estimates to fee_estimates.dat.
```
nit:

```suggestion
// How often to flush fee estimates to disk.
```

willcl-ark commented on src/policy/fees.h:
```diff
@@ -14,12 +14,25 @@
+/** fee_estimates.dat that are more than 60 hours old, 2 and half days will not be read,
+ * as the estimates in the file are stale.
+ */
+static constexpr std::chrono::hours MAX_FILE_AGE{60};
```
nit: Perhaps we can give this a slightly more descriptive name? `FEE_ESTIMATE_MAX_AGE` or something?

Would need to update comment below if you do change

willcl-ark reviewed (commented):
I like the approach here and think persisting mempool min fee is useful. I also agree that re-using fee_estimates.dat would make sense for it, as it likely doesn't need its own file.

Left a few nits and suggestions inline.

ismaelsadeeq committed:
Mempool: add a new property to `MemPoolOptions`

min_relay_feerate_set: indicates whether
`minrelaytxfee` was configured on startup or not.

ismaelsadeeq commented on src/txmempool.cpp:
```diff
@@ -401,8 +404,92 @@ CTxMemPool::CTxMemPool(const Options& opts)
+    }
+    AutoFile min_fee_file{fsbridge::fopen(m_mempool_min_fee_file_path, "rb")};
+
+    // If the file is not found return early
```
no longer there in the recent push

ismaelsadeeq commented on src/txmempool.h:
```diff
@@ -744,6 +744,13 @@ class CTxMemPool
+    /** Flush fee estimates and mempool minimum fee to disk*/
+    void FlushFeeEstimatesAndMempoolMinFee() const EXCLUSIVE_LOCKS_REQUIRED(!cs);
+
+    void SetMemPoolMinFee() const EXCLUSIVE_LOCKS_REQUIRED(!cs);
```
I think it should be considered for tx policy validation 
>While mempool is importing, node receives transactions well below the old mempoolminfee, enters into mempool

See [issue](https://github.com/bitcoin/bitcoin/pull/27622) 

I agree with your point a lot can happen in 60 hours.
I also agree with @willcl-ark  [comment](https://github.com/bitcoin/bitcoin/pull/27622#issuecomment-1592855916)  to make it stricter for reloading `mempoolminfee` only if it's an immediate restart. 

I am taking this to draft while working on that

ismaelsadeeq replied:
I reduced the maximum age of a `mempoolminfee` to one hour.

ismaelsadeeq committed:
Mempool: persist mempool minimum fee accross restart

During shutdown, the mempool minimum fee value is now written to
the 'fee_estimates.dat' file.

On startup, if the user has set the -minrelaytxfee, it will
be used as the mempool minimum fee. Otherwise, if fee_estimates.dat
file is not too old, the maximum value between the persisted mempool min fee
fee and the default mempool minimum fee will be used as the 'rollingMinimumFeeRate'.

The mempool minimum fee will be flushed periodically in case of an
unclean shutdown.

ismaelsadeeq committed:
test: tests persistence of Mempool minimum fee accross restart

Ensure that mempool minimum fee will be flushed to disk
periodically and on shutdown, also tests that stale mempool
minimum fee will not be read.

ismaelsadeeq commented on src/policy/fees.cpp:
```diff
@@ -916,22 +916,23 @@ CFeeRate CBlockPolicyEstimator::estimateSmartFee(int confTarget, FeeCalculation
 }
 
-void CBlockPolicyEstimator::FlushFeeEstimates()
+void CBlockPolicyEstimator::FlushFeeEstimates(const int64_t& mempool_min_fee)
```
Done

BhinbahadurUK commented:
Database re administrator a factor of adding gives values endlessly as once
written data extract comes common to daily living allowing instances acted
or actioned for follow same time and manuotously or read held as if a man
had consecrated auto bind dowload rewipe etc. Basically x factor or its
simply what you make made of it from the start.


Hence my question and assistance how to add developer as such so standards
on my end can be uplifted, remote perhaps and as we have been. Fiat
accounts send to me also with linking.



Thanks and congrats all thus far greatest accomplishments yet .


Go on( see you on the other side🤣)

Regards

On Mon, 19 Jun 2023, 22:44 Abubakar Sadiq Ismail, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/policy/fees.cpp
> <https://github.com/bitcoin/bitcoin/pull/27859#discussion_r1234464010>:
>
> >  }
>
> -void CBlockPolicyEstimator::FlushFeeEstimates()
> +void CBlockPolicyEstimator::FlushFeeEstimates(const int64_t& mempool_min_fee)
>
> Done
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/27859#discussion_r1234464010>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEXF3X2TV2M3E3WYIEAY75LXMC22FANCNFSM6AAAAAAZDDBEVI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

ismaelsadeeq commented on src/txmempool.h:
```diff
@@ -744,6 +744,13 @@ class CTxMemPool
+
+    void SetMemPoolMinFee() const EXCLUSIVE_LOCKS_REQUIRED(!cs);
+
+    int64_t GetMempoolMinFee() const EXCLUSIVE_LOCKS_REQUIRED(!cs);
```
`CFeeRate`'s `nSatoshisPerK` is  `CAmount` same `int64_t`.

ismaelsadeeq replied:
@glozow fixed now using `CAmount`
Thank you

ismaelsadeeq commented:
The issue is after an immediate restart.
- Fee estimates will be remembered but `mempoolminfee` will fall back to `m_incremental_relay_feerate` 1000 sats.
This will make propagating low-fee transactions accepted into the mempool.
- `fee_estimator.estimateSmartFee` will return a non-0 fee rate below 1000 sats (Potentially because of those recently accepted transactions). 
- This makes the `estimatesmartfee` 1000 sats because we are taking `std::max({feeRate, min_mempool_feerate, min_relay_feerate});` 

I also think this happens when low-fee transactions are propagating when node restarts.

After some time the fee estimates will fall back to normal.

So if we persisted `mempoolminfee` and read it only after immediate restarts, We will not accept low-fee transactions that will mess with the fee estimates.

I tested on signet and noticed the estimate after an immediate restart is the same as before.
There aren't low-fee transactions when I restarted.

Except if there is a reason for the `mempoolminfee` to fall back to 1000 sats after an immediate restart, we should not accept these transactions below the `mempoolminfee` to our mempool just because we restart.

ismaelsadeeq commented:
@MarcoFalke Rebased and fix CI failure

ismaelsadeeq commented:
> CI still fails
Should be fixed after this rebase.

I am rethinking the approach taken in this PR, given that there is another PR that decouples the fee estimator from the mempool in progress.
Is it more desirable to persist to `mempool.dat` but that does not flush periodically and will be undesirable to do so. will the initial approach of having its own file be better.

Also conceptually what do you think about this PR Thanks.
@ajtowns @willcl-ark @glozow
